src/widgets/Step.js

```diff
@@ -0,0 +1,158 @@
+ * @prop {String} type  - The type of step, select (dropdown) or input (text field).
+ * @prop {String} stepKey - A string key for this step.
+ * @prop {String} status - The status of this step (as above)
+ * @prop {Func}   getModalData - Callback for fetching this steps modal data
+ * @prop {Func}   onPress      - Callback for pressing this step (opens modal)
+ * @state modalData - Data to pass to the modal on opening and for determining if the
+ * current step is in an error state. Fetched when the status of this step changes,
+ * and stored for performance as to not re-fetch every re-render
+ */
+export const Step = memo(props => {
+  const [modalData, setModalData] = useState([]);
+  const { data, field, type, stepKey, status, getModalData, onPress } = props;
+  // Each stepKey used with this component has a placeholder/error string in getLocalisation.
+  const { placeholder, errorString } = getLocalisation({ stepKey });
+  // Calculate the errorState in this component after fetching modalData. If there is no
+  // modalData (there are no options to choose from), display the error string and a new icon.
+  const errorState = !(modalData && modalData.length) && status === 'CURRENT' && getModalData;
+  const Container = errorState || status === 'INCOMPLETE' ? View : TouchableOpacity;
+  // Fetch modalData when the status is current and store it in state.
+  useEffect(() => {
+    if (status !== 'CURRENT' || !getModalData) return;
+    setModalData(getModalData());
+  }, [status, errorState]);
+
+  const onSelection = () => {
+    onPress({ selection: modalData, key: stepKey });
+  };
+
+  /** Inner components */
+
+  // Render a red X on error state, nothing on incomplete status, a green tick when
+  // complete or a orange arrow on the current step.
+  const StepIcon = () => {
+    const { iconContainerStyle } = localStyles;
+    let icon;
+    if (errorState) icon = <Ionicons name="ios-close" color="red" size={30} />;
+    else if (status === 'INCOMPLETE') icon = null;
+    else if (status === 'COMPLETE') icon = <Ionicons name="md-checkmark" color="green" size={30} />;
+    else if (status === 'CURRENT') {
+      icon = <Ionicons name="md-arrow-round-forward" color={SUSSOL_ORANGE} size={30} />;
+    }
+
+    return <View style={iconContainerStyle}>{icon}</View>;
+  };
+
+  // Render a pencil on text editing or a down arrow for list selections.
+  const EditIcon = () => {
+    if (errorState) return null;
+    return (
+      <View>
+        {type === 'input' && <FontAwesome name="pencil" size={20} color="white" />}
+        {type !== 'input' && <Ionicons name="ios-arrow-down" size={20} color="white" />}
+      </View>
+    );
+  };
+
+  // Render the error string if in an error state, the title if there is
+  // data/the step is complete and a placeholder otherwise.
+  const TextDisplay = () => {
+    const { textContainerStyle, textStyle } = localStyles;
+    const text = (data && data[field]) || data;
```

Comment:
Maybe a target for future refactoring to make data always exclusively either `string` or `object`?

Reply:
I suppose so, makes the code ugly but is a bit easier to be a little flexible